node/pkg/processor/reactor/reactor.go

```diff
@@ -0,0 +1,491 @@
+			zap.Error(err),
+			zap.String("digest", hash),
+			zap.String("signature", hex.EncodeToString(m.Signature)),
+			zap.String("addr", hex.EncodeToString(m.Addr)),
+		)
+		return
+	}
+	theirAddr := ethcommon.BytesToAddress(m.Addr)
+	observationsReceivedByGuardianAddressTotal.WithLabelValues(c.group, theirAddr.Hex()).Inc()
+
+	c.logger.Debug("accepted foreign observation", zap.Any("observation", m), zap.Stringer("address", theirAddr))
+
+	// Have we already received this observation
+	if _, has := c.signatures[theirAddr]; has {
+		// TODO log duplicate
+		return
+	}
+
+	// Store their signature
+	c.signatures[theirAddr] = m.Signature
+	c.lastObservation = time.Now()
+
+	if c.state == StateInitialized {
+		c.logger.Debug("received observation before own observation", zap.Any("observation", m))
+		c.firstSeen = time.Now()
+		c.stateTransition(StateUnobserved)
+	}
+
+	// If we haven't reached quorum yet, there is nothing more to do
+	if !c.hasQuorum() {
+		return
+	}
+
+	// Transition to quorum states
+	switch c.state {
+	case StateObserved:
+		reactorQuorum.WithLabelValues(c.group, "quorum").Inc()
+		c.stateTransition(StateQuorum)
+	case StateUnobserved:
+		reactorQuorum.WithLabelValues(c.group, "quorum_unobserved").Inc()
+		c.stateTransition(StateQuorumUnobserved)
+	}
+}
+
+func (c *ConsensusReactor[K]) observed(ctx context.Context, o K) {
+	c.stateMutex.Lock()
+	defer c.stateMutex.Unlock()
+
+	c.logger.Debug("observed", zap.Any("observation", o))
+
+	if c.state != StateInitialized && c.state != StateUnobserved && c.state != StateQuorumUnobserved {
+		return
+	}
+
+	// Late observation
+	if c.state == StateQuorumUnobserved {
+		reactorObservedLate.WithLabelValues(c.group).Inc()
+	}
+
+	c.observation = o
+
+	if c.config.Signer != nil {
+		// Generate digest of the unsigned VAA.
+		digest := o.SigningMsg()
+
+		// Sign the digest using our node's guardian key.
+		timeout, cancel := context.WithTimeout(ctx, time.Second*5)
+		defer cancel()
+		localAddr, err := c.config.Signer.Address(timeout)
+		if err != nil {
+			panic(err)
+		}
+		s, err := c.config.Signer.Sign(timeout, digest.Bytes())
+		if err != nil {
+			panic(err)
+		}
+		c.localSignature = s
+
+		messagesSignedTotal.WithLabelValues(c.group).Inc()
+
+		// Store in signatures array
+		c.signatures[localAddr] = s
+
+		err = c.transmitSignature(ctx)
+		if err != nil {
+			c.logger.Error("failed to transmit signature on observation", zap.Error(err))
+		}
+
+	}
+
+	// Transition to quorum states
+	switch c.state {
+	case StateInitialized:
+		c.firstSeen = time.Now()
+		c.lastObservation = time.Now()
+		c.stateTransition(StateObserved)
+	case StateUnobserved:
+		c.lastObservation = time.Now()
+		c.stateTransition(StateObserved)
+	case StateQuorumUnobserved:
+		reactorQuorum.WithLabelValues(c.group, "quorum").Inc()
+		c.stateTransition(StateQuorum)
+	}
+
+	// If we haven't reached quorum in this event, there is nothing more to do
+	if !c.hasQuorum() && c.state != StateQuorum {
```

Comment:
This is supposed to prevent the state transition from being triggered twice in the case that L409 transitioned to `StateQuorum` already. 

I'll make this more clear by just adding a return there.